File: libs/panoalgo/tests/test_limits.py
"""L0 mutlak limit testleri — PLAN.md TA2 Adim 5.

Beklenen degerlerin kaynagi:
  - TUM esikler contracts/alarm-codes.yaml'dan `thresholds` fixture'i ile okunur;
    bu dosyada 50 / 70 / 105 / 15 / 2312 gibi HICBIR sayi elle yazili degildir
    (PLAN.md kural 10).
  - Karsilastirma yonu backend/app/risk.py:58-59 ile ayni olmak ZORUNDA: kesin
    buyuktur (`>`). backend/tests/test_risk.py tam esikteki degerin alarm
    URETMEDIGINI kilitliyor; panoalgo `>=` kullanirsa alarm konsolu kirmizi
    gosterirken dijital ikizde nokta yesil kalir.
  - Faz farki gruplamasi ve notr dislamasi da risk.py:38 `^(GIRIS|DSYA\\d)_L[123]$`
    deseniyle ayni.

Ornek yuk, TA1 uretecinden alinir: boylece testler gercek sema-gecerli bir yuk
uzerinde calisir, elle kurulmus yapay sozluk uzerinde degil.
"""

from __future__ import annotations

import pytest

from panoalgo.limits import evaluate


def _set_point(sample: dict, pt: str, **fields) -> dict:
    for point in sample["t_conn"]:
        if point["pt"] == pt:
            point.update(fields)
            return sample
    raise KeyError(pt)


def _flatten_rises(sample: dict, value: float) -> dict:
    for point in sample["t_conn"]:
        point["dt_c"] = value
    return sample


# --------------------------------------------------------------- saglikli taban


def test_healthy_sample_raises_no_l0_alarm(sample):
    assert evaluate(sample) == []


def test_result_contains_only_codes_defined_in_the_contract(sample, alarm_codes):
    known = {a["code"] for a in alarm_codes["alarms"]}
    _set_point(sample, "GIRIS_L1", dt_c=999.0)
    sample["env"]["td_margin_k"] = -5.0
    sample["elec"]["i_ph"] = [9999.0, 9999.0, 9999.0]

    assert set(evaluate(sample)) <= known


# ------------------------------------------------------------ sicaklik artisi


def test_terminal_warning_fires_just_above_the_threshold(sample, thresholds):
    _set_point(sample, "DSYA3_L2", dt_c=thresholds["term_rise_warn_k"] + 0.1)
    assert "ALM-THR-TERM-WARN" in evaluate(sample)


def test_terminal_warning_does_not_fire_exactly_at_the_threshold(sample, thresholds):
    """risk.py:58 `value > limit` — tam esik alarm DEGILDIR."""
    _flatten_rises(sample, thresholds["term_rise_warn_k"])
    assert "ALM-THR-TERM-WARN" not in evaluate(sample)


def test_terminal_alarm_fires_above_seventy_kelvin(sample, thresholds):
    _set_point(sample, "GIRIS_L2", dt_c=thresholds["term_rise_alarm_k"] + 1.0)
    codes = evaluate(sample)
    assert "ALM-THR-TERM-ALM" in codes
    assert "ALM-THR-TERM-WARN" in codes, "70 K'yi gecen deger 50 K'yi de gecmistir"


def test_busbar_alarm_fires_above_its_own_threshold(sample, thresholds):
    _set_point(sample, "GIRIS_L1", dt_c=thresholds["bus_rise_alarm_k"] + 1.0)
    assert "ALM-THR-BUS-ALM" in evaluate(sample)


def test_busbar_alarm_stays_silent_between_terminal_and_busbar_limits(sample, thresholds):
    _set_point(sample, "GIRIS_L1", dt_c=thresholds["term_rise_alarm_k"] + 1.0)
    assert "ALM-THR-BUS-ALM" not in evaluate(sample)


# --------------------------------------------------------------- faz farki


def test_phase_difference_alarm_uses_the_hottest_and_coldest_phase_of_a_group(sample, thresholds):
    _flatten_rises(sample, 10.0)
    _set_point(sample, "DSYA4_L1", dt_c=10.0 + thresholds["phase_diff_alarm_k"] + 1.0)
    assert "ALM-THR-PHASE-DIF" in evaluate(sample)


def test_phase_difference_ignores_the_neutral_point(sample, thresholds):
    """risk.py:38 notru desen disinda birakir; notr fazlardan soguk olmasi normaldir."""
    _flatten_rises(sample, 10.0 + thresholds["phase_diff_alarm_k"] + 1.0)
    _set_point(sample, "GIRIS_N", dt_c=0.0)
    assert "ALM-THR-PHASE-DIF" not in evaluate(sample)


def test_phase_difference_does_not_compare_across_different_feeders(sample, thresholds):
    """DSYA1 sicak, DSYA5 soguk olabilir — farkli fiderler farkli yuk tasir."""
    _flatten_rises(sample, 5.0)
    for pt in ("DSYA1_L1", "DSYA1_L2", "DSYA1_L3"):
        _set_point(sample, pt, dt_c=5.0 + thresholds["phase_diff_alarm_k"] + 2.0)
    assert "ALM-THR-PHASE-DIF" not in evaluate(sample)


# ------------------------------------------------------------------- akim


def test_overcurrent_uses_the_main_input_rating(sample, thresholds):
    """risk.py:198-203 tek limit kullanir: current_alarm_ratio * rated main_input."""
    limit = thresholds["current_alarm_ratio"] * thresholds["rated_current_a"]["main_input"]
    sample["elec"]["i_ph"] = [limit + 1.0, 10.0, 10.0]
    assert "ALM-I-OVER" in evaluate(sample)


def test_overcurrent_does_not_fire_exactly_at_the_rating(sample, thresholds):
    limit = thresholds["current_alarm_ratio"] * thresholds["rated_current_a"]["main_input"]
    sample["elec"]["i_ph"] = [limit, limit, limit]
    assert "ALM-I-OVER" not in evaluate(sample)


# ---------------------------------------------------------------- yogusma


def test_dew_warning_fires_below_its_threshold(sample, thresholds):
    sample["env"]["td_margin_k"] = thresholds["dew_margin_warn_k"] - 0.1
    codes = evaluate(sample)
    assert "ALM-DEW-WARN" in codes
    assert "ALM-DEW-ALM" not in codes


def test_dew_alarm_fires_when_condensation_is_imminent(sample, thresholds):
    sample["env"]["td_margin_k"] = thresholds["dew_margin_alarm_k"] - 0.1
    codes = evaluate(sample)
    assert "ALM-DEW-ALM" in codes
    assert "ALM-DEW-WARN" in codes, "1 K'nin altindaki marj 3 K'nin da altindadir"


def test_dew_warning_silent_exactly_at_the_threshold(sample, thresholds):
    sample["env"]["td_margin_k"] = thresholds["dew_margin_warn_k"]
    assert "ALM-DEW-WARN" not in evaluate(sample)


# ------------------------------------------------------------- pano ortami


def test_panel_temperature_alarm_uses_the_upper_air_node(sample, thresholds):
    sample["env"]["t_up_c"] = thresholds["panel_temp_alarm_c"] + 1.0
    assert "ALM-PANEL-TEMP" in evaluate(sample)


# --------------------------------------------------------- koruma sagligi


def test_protection_health_alarm_when_the_arc_guard_reports_a_fault(sample):
    sample["tvoc"]["prot_health_ok"] = False
    assert "ALM-PROT-HEALTH" in evaluate(sample)


def test_protection_health_alarm_when_the_arc_guard_stops_answering(sample):
    """Cevap vermeyen TVOC-2 = pano sessizce korumasiz (rapor 3.5, fabrika ID 248)."""
    sample["tvoc"]["comm_ok"] = False
    assert "ALM-PROT-HEALTH" in evaluate(sample)


def test_no_protection_alarm_when_the_panel_has_no_arc_guard(sample):
    """tvoc null olabilir (sema); bu 'koruma arizali' demek DEGILDIR."""
    sample["tvoc"] = None
    assert "ALM-PROT-HEALTH" not in evaluate(sample)


# ------------------------------------------------------------------- ark


def test_arc_trip_needs_the_trip_counter_to_change(sample):
    """PDU 149 sayaci ARTARSA yeni ark olayi. Sayac sifirlanmadigi icin
    'trips > 0' kurali alarmi sonsuza kadar mandallardi."""
    previous = {"tvoc": {"trips": 3}}
    sample["tvoc"]["trips"] = 4
    assert "ALM-ARC-TRIP" in evaluate(sample, previous)


def test_arc_trip_does_not_latch_on_a_historic_trip_count(sample):
    previous = {"tvoc": {"trips": 4}}
    sample["tvoc"]["trips"] = 4
    assert "ALM-ARC-TRIP" not in evaluate(sample, previous)


def test_arc_trip_is_not_reported_without_a_previous_sample(sample):
    sample["tvoc"]["trips"] = 4
    assert "ALM-ARC-TRIP" not in evaluate(sample)


# ------------------------------------------------------------- dayaniklilik


def test_evaluate_never_returns_the_centre_owned_comms_alarm(sample):
    """ALM-COMMS-LOST merkezin kendi kodudur (backend/app/alarm_service.py);
    kenardan da donerse iki mekanizma ayni alarmi birbirine karsi yonetir."""
    sample["health"]["nodes_ok"] = 0
    assert "ALM-COMMS-LOST" not in evaluate(sample)


def test_evaluate_returns_no_duplicate_codes(sample, thresholds):
    _flatten_rises(sample, thresholds["term_rise_alarm_k"] + 1.0)
    codes = evaluate(sample)
    assert len(codes) == len(set(codes))


def test_evaluate_tolerates_a_payload_with_only_the_required_fields(sample):
    """Opsiyonel bloklar (tvoc, pd, risk) sema geregi eksik olabilir; patlamamali."""
    for optional in ("tvoc", "pd", "risk", "alarms", "fw"):
        sample.pop(optional, None)
    assert isinstance(evaluate(sample), list)


@pytest.mark.parametrize("missing", ["elec", "env", "t_conn"])
def test_evaluate_does_not_raise_on_a_malformed_payload(sample, missing):
    """B'nin risk.py:95 detect() cagrisini try/except ile sarmiyor — bir istisna
    TUM ingest partisini dusururdu. Bozuk yukte bos liste donmek zorundayiz."""
    sample.pop(missing)
    assert evaluate(sample) == []


# ------------------------------------- K indeksi ve sinira kalan sure (L1 esikleri)
#
# Bu kodlar sozlesmede L1 etiketlidir, ama KARARLARI yine duz esik karsilastirmasidir
# (k_ratio ve ttl_h alanlari zaten yukte hazir gelir). Uretim yeri detect.py, KARAR
# yeri burasi — backend/app/risk.py:29-36 POINT_LIMITS tablosu da ayni ayrimi yapar.


def test_k_ratio_warning_fires_above_its_threshold(sample, thresholds):
    _set_point(sample, "DSYA2_L1", k_ratio=thresholds["k_ratio_warn"] + 0.01)
    codes = evaluate(sample)
    assert "ALM-K-WARN" in codes
    assert "ALM-K-ALM" not in codes


def test_k_ratio_alarm_fires_above_its_threshold(sample, thresholds):
    _set_point(sample, "DSYA2_L1", k_ratio=thresholds["k_ratio_alarm"] + 0.01)
    codes = evaluate(sample)
    assert "ALM-K-ALM" in codes
    assert "ALM-K-WARN" in codes, "1.6'yi gecen oran 1.3'u de gecmistir"


def test_k_ratio_silent_exactly_at_the_threshold(sample, thresholds):
    _set_point(sample, "DSYA2_L1", k_ratio=thresholds["k_ratio_warn"])
    assert "ALM-K-WARN" not in evaluate(sample)


def test_healthy_k_ratio_of_one_raises_nothing(sample):
    assert "ALM-K-WARN" not in evaluate(sample)


def test_time_to_limit_alarm_uses_hours_not_days(sample, thresholds):
    """Sozlesme esigi GUN (ttl_warn_days=14), yuk alani SAAT (ttl_h).
    Cevrim unutulursa 14 saat kala degil 14 GUN kala alarm verilir."""
    hours = thresholds["ttl_warn_days"] * 24.0
    _set_point(sample, "DSYA5_L3", ttl_h=hours - 1.0)
    assert "ALM-TTL-14D" in evaluate(sample)


def test_time_to_limit_silent_when_there_is_more_time_than_the_threshold(sample, thresholds):
    hours = thresholds["ttl_warn_days"] * 24.0
    _set_point(sample, "DSYA5_L3", ttl_h=hours + 1.0)
    assert "ALM-TTL-14D" not in evaluate(sample)


def test_time_to_limit_ignores_points_without_an_estimate(sample):
    """ttl_h = null 'tahmin yok' demektir, 'sifir saat kaldi' demek DEGILDIR."""
    for point in sample["t_conn"]:
        point["ttl_h"] = None
    assert "ALM-TTL-14D" not in evaluate(sample)


def test_codes_are_returned_in_contract_bit_order(sample, thresholds, alarm_codes):
    """Sirali cikti, testleri ve gunluk kayitlari kararli kilar."""
    _flatten_rises(sample, thresholds["bus_rise_alarm_k"] + 1.0)
    sample["env"]["td_margin_k"] = -1.0

    bits = {a["code"]: a["bit"] for a in alarm_codes["alarms"]}
    codes = evaluate(sample)
    assert codes == sorted(codes, key=lambda c: bits[c])

File: firmware/tests/test_limits.c
/* L0/L1 esik karari testi (TA3, Kisi A).
 *
 * Python tarafindaki libs/panoalgo/tests/test_limits.py ile AYNI davranislari
 * kilitler. Esikler sozlesme degerleridir (contracts/alarm-codes.yaml) ve teste
 * konfigurasyon olarak verilir — gomulu tarafta da oyle gelecek.
 *
 * En kritik iki iddia:
 *   - tam esikteki deger alarm URETMEZ (kesin buyuktur),
 *   - faz dengesizligi TEK BASINA faz farki alarmi uretmez ("benzer yukte" kosulu).
 */
#include "limits.h"

#include <stdio.h>

#define BIT(n) (1UL << (unsigned)(n))

/* contracts/alarm-codes.yaml thresholds blogundan. */
static pano_limits_t contract_limits(void)
{
    pano_limits_t lim;
    lim.term_rise_warn_k = 50.0;
    lim.term_rise_alarm_k = 70.0;
    lim.bus_rise_alarm_k = 105.0;
    lim.phase_diff_alarm_k = 15.0;
    lim.k_ratio_warn = 1.3;
    lim.k_ratio_alarm = 1.6;
    lim.ttl_warn_h = 14.0 * 24.0;      /* sozlesme GUN verir, burada SAAT */
    lim.dew_margin_warn_k = 3.0;
    lim.dew_margin_alarm_k = 1.0;
    lim.panel_temp_alarm_c = 45.0;
    lim.current_alarm_ratio = 1.00;
    lim.rated_current_a = 2312.0;
    lim.similar_load_max_ratio = 1.10;
    return lim;
}

/* Saglikli bir pano: 3 faz GIRIS + 3 faz DSYA1, hepsi serin ve dengeli. */
static pano_sample_t healthy_sample(pano_point_t *points)
{
    for (int i = 0; i < 6; ++i) {
        points[i].dt_c = 20.0;
        points[i].k_ratio = 1.0;
        points[i].ttl_h = 0.0;
        points[i].has_ttl = 0;
        points[i].group = (i < 3) ? 0 : 1;
        points[i].phase = i % 3;
    }

    pano_sample_t s;
    s.points = points;
    s.point_count = 6;
    s.i_ph[0] = 1000.0;
    s.i_ph[1] = 1000.0;
    s.i_ph[2] = 1000.0;
    s.td_margin_k = 8.0;
    s.t_up_c = 38.0;
    s.tvoc_present = 1;
    s.prot_health_ok = 1;
    s.comm_ok = 1;
    s.trips = 4;
    s.prev_trips = 4;
    s.has_prev_trips = 1;
    return s;
}

static int checks = 0;
static int failures = 0;

static void expect(int condition, const char *what)
{
    checks++;
    if (!condition) {
        fprintf(stderr, "BASARISIZ: %s\n", what);
        failures++;
    }
}

int main(void)
{
    const pano_limits_t lim = contract_limits();
    pano_point_t points[6];

    /* --- saglikli taban --- */
    pano_sample_t s = healthy_sample(points);
    expect(pano_limits_evaluate(&lim, &s) == 0UL, "saglikli panoda hicbir alarm cikmamali");

    /* --- sicaklik artisi: tam esik alarm DEGIL --- */
    s = healthy_sample(points);
    for (int i = 0; i < 6; ++i) { points[i].dt_c = lim.term_rise_warn_k; }
    expect(!(pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_THR_TERM_WARN)),
           "tam esikteki artis uyari uretmemeli (kesin buyuktur)");

    points[2].dt_c = lim.term_rise_warn_k + 0.1;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_THR_TERM_WARN)) != 0UL,
           "esigin hemen ustu uyari uretmeli");

    /* --- 70 K'yi gecen deger 50 K'yi de gecmistir --- */
    s = healthy_sample(points);
    points[0].dt_c = lim.term_rise_alarm_k + 1.0;
    {
        const unsigned long bits = pano_limits_evaluate(&lim, &s);
        expect((bits & BIT(PANO_BIT_THR_TERM_ALM)) != 0UL, "70 K ustu alarm uretmeli");
        expect((bits & BIT(PANO_BIT_THR_TERM_WARN)) != 0UL, "70 K ustu uyariyi da icermeli");
        expect((bits & BIT(PANO_BIT_THR_BUS_ALM)) == 0UL, "105 K altinda bara alarmi olmamali");
    }

    /* --- faz farki: BENZER YUKTE --- */
    s = healthy_sample(points);
    points[3].dt_c = 20.0 + lim.phase_diff_alarm_k + 1.0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_THR_PHASE_DIF)) != 0UL,
           "benzer yukte 15 K ustu fark alarm uretmeli");

    /* Ayni sicaklik farki, ama fazlar farkli yuk tasiyor -> anomali DEGIL. */
    s.i_ph[0] = 1000.0;
    s.i_ph[1] = 1000.0;
    s.i_ph[2] = 1400.0;   /* oran 1.4 > similar_load_max_ratio */
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_THR_PHASE_DIF)) == 0UL,
           "dengesiz yukte ham sicaklik farki alarm uretmemeli");

    /* Notr faz karsilastirmasina girmez. */
    s = healthy_sample(points);
    points[0].phase = -1;
    points[0].dt_c = 0.0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_THR_PHASE_DIF)) == 0UL,
           "notr nokta faz farkina girmemeli");

    /* Farkli fiderler birbiriyle kiyaslanmaz. */
    s = healthy_sample(points);
    for (int i = 3; i < 6; ++i) { points[i].dt_c = 20.0 + lim.phase_diff_alarm_k + 2.0; }
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_THR_PHASE_DIF)) == 0UL,
           "farkli fiderler arasi fark alarm uretmemeli");

    /* --- K/K0 --- */
    s = healthy_sample(points);
    points[4].k_ratio = lim.k_ratio_warn;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_K_WARN)) == 0UL,
           "tam esikteki K/K0 uyari uretmemeli");
    points[4].k_ratio = lim.k_ratio_alarm + 0.01;
    {
        const unsigned long bits = pano_limits_evaluate(&lim, &s);
        expect((bits & BIT(PANO_BIT_K_ALM)) != 0UL, "1.6 ustu K/K0 alarm uretmeli");
        expect((bits & BIT(PANO_BIT_K_WARN)) != 0UL, "1.6 ustu 1.3'u de gecmistir");
    }

    /* --- sinira kalan sure: SAAT, ve "tahmin yok" != "sifir saat" --- */
    s = healthy_sample(points);
    points[1].has_ttl = 1;
    points[1].ttl_h = lim.ttl_warn_h - 1.0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_TTL_14D)) != 0UL,
           "14 gunden az kalan sure alarm uretmeli");
    points[1].ttl_h = lim.ttl_warn_h + 1.0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_TTL_14D)) == 0UL,
           "14 gunden fazla kalan sure alarm uretmemeli");
    s = healthy_sample(points);   /* has_ttl = 0 */
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_TTL_14D)) == 0UL,
           "tahmin yokken ttl alarmi cikmamali");

    /* --- asiri akim: ana giris anma degeri --- */
    s = healthy_sample(points);
    s.i_ph[0] = lim.current_alarm_ratio * lim.rated_current_a;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_I_OVER)) == 0UL,
           "tam anma akimi alarm uretmemeli");
    s.i_ph[0] = lim.current_alarm_ratio * lim.rated_current_a + 1.0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_I_OVER)) != 0UL,
           "anma akiminin ustu alarm uretmeli");

    /* --- yogusma --- */
    s = healthy_sample(points);
    s.td_margin_k = lim.dew_margin_warn_k - 0.1;
    {
        const unsigned long bits = pano_limits_evaluate(&lim, &s);
        expect((bits & BIT(PANO_BIT_DEW_WARN)) != 0UL, "3 K altinda yogusma uyarisi");
        expect((bits & BIT(PANO_BIT_DEW_ALM)) == 0UL, "3 K altinda henuz alarm yok");
    }
    s.td_margin_k = lim.dew_margin_alarm_k - 0.1;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_DEW_ALM)) != 0UL,
           "1 K altinda yogusma alarmi");

    /* --- pano ic sicakligi --- */
    s = healthy_sample(points);
    s.t_up_c = lim.panel_temp_alarm_c + 1.0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_PANEL_TEMP)) != 0UL,
           "45 degC ustu pano sicaklik alarmi");

    /* --- koruma sagligi ve ark --- */
    s = healthy_sample(points);
    s.prot_health_ok = 0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_PROT_HEALTH)) != 0UL,
           "dedektor arizasi koruma alarmi uretmeli");

    s = healthy_sample(points);
    s.comm_ok = 0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_PROT_HEALTH)) != 0UL,
           "cevap vermeyen TVOC-2 koruma alarmi uretmeli");

    s = healthy_sample(points);
    s.tvoc_present = 0;
    s.prot_health_ok = 0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_PROT_HEALTH)) == 0UL,
           "TVOC-2 olmayan panoda koruma alarmi cikmamali");

    s = healthy_sample(points);
    s.trips = 5;   /* prev_trips = 4 */
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_ARC_TRIP)) != 0UL,
           "artan trip sayaci ark alarmi uretmeli");

    s = healthy_sample(points);
    s.trips = 9;
    s.prev_trips = 9;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_ARC_TRIP)) == 0UL,
           "degismeyen trip sayaci alarmi mandallamamali");

    s = healthy_sample(points);
    s.trips = 9;
    s.has_prev_trips = 0;
    expect((pano_limits_evaluate(&lim, &s) & BIT(PANO_BIT_ARC_TRIP)) == 0UL,
           "onceki ornek yokken gecis karari verilmemeli");

    if (failures > 0) {
        fprintf(stderr, "%d/%d kontrol basarisiz\n", failures, checks);
        return 1;
    }
    printf("%d kontrol gecti (esikler sozlesmeden, karsilastirma kesin buyuktur).\n", checks);
    return 0;
}
